src/osd/PG.cc

```diff
@@ -2720,49 +2739,167 @@ void PG::_update_calc_stats()
     // Objects missing from nodes not in up, sort by # objects
     boost::container::flat_set<pair<int64_t,pg_shard_t>> acting_source_objects;
 
-    int64_t missing;
+    // This is used in recovery to use the actual location of things
+    // to determine degraded or misplaced
+    map<hobject_t, set< pg_shard_t > > missing_objects;
```

Comment:
If i'm reading this right, this map is generated for all missing objects each time we call calc_update_stats(). I think that's a lot--maybe every IO on the PG even?

I think we can get the same information by extending MissingLoc in PG.h a bit.  Instead of just missing_loc_sources like so
```
    map<hobject_t, set<pg_shard_t> > missing_loc;
    set<pg_shard_t> missing_loc_sources;
```
we can instead do
```
    map<pg_shard_t,int> missing_loc_sources;
```
where the value is the number of missing objects present on that source.  This is a private member of MissingLoc, so there should only be a few places where we need to update it to keep in sync with missing_loc itself.. it'd be the number of times the shard appears in missing_loc.

Is that then enough info to do the more accurate missing calculate below?

Reply:
The missing_objects map is only filled in when (is_remapped() || is_undersized() || !is_clean()) && (is_peered() || is_activating()) AND when recovering which includes missing information.

Reply:
True, but during that log recovery period it's a big impact.  If I'm not mistaken, changing hte MissingLoc tracking just slightly should have almost no net impact and should give us the information needed, right?

Reply:
@liewegas The missing_loc_by_shard is not enough information.  We have to find degraded objects reduced by any extra objects.  All extra objects not used to reduce degraded are misplaced.

Reply:
Yeah you're right, was I misremembering.  

What about this:
```
map<int,int> missing_by_num_loc;
````
which maps the number of targets (size of missing_loc[whatever]) to the number of missing objects with that many locations.

(Or, perhaps something similar, but broken down so that is it only summing for locations that aren't part of acting?)

Reply:
Or I guess ``upset``.  In any case, missing_loc is reset on each peering interval, so we could give it a copy of upset so that it could manage these counts as it is constructed and updated after recoveries complete.